hack/generator/pkg/astmodel/storage_conversion_factories.go

```diff
@@ -551,186 +488,41 @@ func assignEnumTypeFromOptionalEnumType(
-
-	return func(reader dst.Expr, writer dst.Expr, ctx *CodeGenerationContext) []dst.Stmt {
-		// Need to check for null and only assign if we have a value
-		cond := astbuilder.NotEqual(reader, dst.NewIdent("nil"))
-
-		readValue := astbuilder.SimpleAssignment(
-			dst.NewIdent(local),
-			token.DEFINE,
-			astbuilder.CallFunc(dstName.name, astbuilder.Dereference(reader)))
-
-		writeValue := astbuilder.SimpleAssignment(
-			writer,
-			token.ASSIGN,
-			dst.NewIdent(local))
+	local := destinationEndpoint.CreateLocal("", "Enum")
 
-		assignZero := astbuilder.SimpleAssignment(
-			writer,
-			token.ASSIGN,
-			astbuilder.CallFunc(dstName.name,
-				&dst.BasicLit{
-					Value: zeroValue(srcEnum.baseType),
-				}))
+	return func(reader dst.Expr, writer func(dst.Expr) []dst.Stmt, ctx *CodeGenerationContext) []dst.Stmt {
 
-		stmt := &dst.IfStmt{
-			Cond: cond,
-			Body: astbuilder.StatementBlock(readValue, writeValue),
-			Else: astbuilder.StatementBlock(assignZero),
-		}
+		declaration := astbuilder.LocalVariableDeclaration(local, dst.NewIdent(dstName.Name()), "")
 
-		return []dst.Stmt{stmt}
-	}
-}
-
-// assignOptionalEnumTypeFromEnumType will generate a direct assignment if both types have the same
-// underlying primitive type and only the destination is optional
-//
-// <destination> = <enum>(<source>)
-//
-func assignOptionalEnumTypeFromEnumType(
-	sourceEndpoint *StorageConversionEndpoint,
-	destinationEndpoint *StorageConversionEndpoint,
-	conversionContext *StorageConversionContext) StorageTypeConversion {
-
-	// Require source to be non-optional
-	if _, srcOpt := AsOptionalType(sourceEndpoint.Type()); srcOpt {
-		return nil
-	}
-
-	// Require destination to be optional
-	if _, dstOpt := AsOptionalType(destinationEndpoint.Type()); !dstOpt {
-		return nil
-	}
-
-	// Require source to be an enumeration
-	_, srcType, ok := conversionContext.ResolveType(sourceEndpoint.Type())
-	if !ok {
-		return nil
-	}
-	srcEnum, srcIsEnum := AsEnumType(srcType)
-	if !srcIsEnum {
-		return nil
-	}
-
-	// Require destination to be an enumeration
-	dstName, dstType, ok := conversionContext.ResolveType(destinationEndpoint.Type())
-	if !ok {
-		return nil
-	}
-	dstEnum, dstIsEnum := AsEnumType(dstType)
-	if !dstIsEnum {
-		return nil
-	}
-
-	// Require both enumerations to have the same underlying type
-	if !srcEnum.baseType.Equals(dstEnum.baseType) {
-		return nil
-	}
-
-	copyVar := destinationEndpoint.CreateLocal("Value")
-
-	return func(reader dst.Expr, writer dst.Expr, ctx *CodeGenerationContext) []dst.Stmt {
-		return []dst.Stmt{
-			astbuilder.SimpleAssignment(
-				dst.NewIdent(copyVar),
-				token.DEFINE,
-				astbuilder.CallFunc(dstName.name, reader)),
+		// Need to check for null and only assign if we have a value
+		cond := astbuilder.NotEqual(reader, dst.NewIdent("nil"))
 
-			astbuilder.SimpleAssignment(
-				writer,
-				token.ASSIGN,
-				astbuilder.AddrOf(dst.NewIdent(copyVar))),
+		writeLocal := func(expr dst.Expr) []dst.Stmt {
```

Comment:
I think this is the same as the other writeLocal, isn't it?

Reply:
Yes, but again it needs to be a local func because of the constrained signature.